### PR TITLE
NP-615: Live migration

### PR DESCRIPTION
**Pull request by pliurh**
Enhancement: https://github.com/openshift/enhancements/pull/1064

### CODE REVIEW

**Comment by pliurh** [comment]
This PR is blocked https://github.com/openshift/api/pull/1640. Besides the API update, it is ready for review.
/cc @kyrtapz @martinkennelly

File: bindata/network/openshift-sdn/controller.yaml [hunk 1]
```diff
@@ -27,6 +27,10 @@ spec:
         - /bin/bash
         - -c
         - |
+{{- if .IsNetworkTypeMigration }}
+          echo "wait for hostsubnets in live migration mode"
+          while ! oc get node --no-headers| awk '{ print $1 }'| xargs -i oc get hostsubnet {}; do sleep 3; done
```
Comment (kyrtapz) [inline bindata/network/openshift-sdn/controller.yaml:32]: NIT: to get just the node names you can do:
```
oc get nodes -o custom-columns=NAME:.metadata.name --no-headers
```

File: bindata/network/openshift-sdn/sdn.yaml [hunk 2]
```diff
@@ -66,7 +66,56 @@ spec:
         - |
           #!/bin/bash
           set -euo pipefail
+{{- if .IsNetworkTypeMigration }}
+          set +e
 
+          SDN_SUBNET=$(oc get hostsubnet ${K8S_NODE_NAME} -o jsonpath="{.subnet}")
+          echo "sdn hostsubnet:" ${SDN_SUBNET}
+          OVN_SUBNET=$(oc get node ${K8S_NODE_NAME} -o jsonpath='{.metadata.annotations.k8s\.ovn\.org/node-subnets}'| grep -o  -E '[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\/[0-9]{1,3}')
```
Comment (kyrtapz) [inline bindata/network/openshift-sdn/sdn.yaml:74]: Would it be possible to use `jq '.default[0]'` instead of the grep?
I think the grep might not work if there are multiple node-subnets (secondary networks)

File: bindata/network/openshift-sdn/sdn.yaml [hunk 3]
```diff
@@ -66,7 +66,56 @@ spec:
...
+          if [ ! -z "${SDN_SUBNET}" ]; then
+              echo "sdn hostsubnet is set"
+              SUBNET=${SDN_SUBNET}
+          else
+              echo "sdn hostsubnet is unset"
+              if [ ! -z "${OVN_SUBNET}" ]; then
+                  echo "hostsubnet CR is not found, use the value from ovnkube annotation"
+                  SUBNET=${OVN_SUBNET}
+                  NODE_UID=$(oc get node ${K8S_NODE_NAME} -o jsonpath='{.metadata.uid}')
+                  NODE_IP=$(oc get nodes ${K8S_NODE_NAME} -o jsonpath='{.status.addresses[?(@.type=="InternalIP")].address}')
```
Comment (kyrtapz) [inline bindata/network/openshift-sdn/sdn.yaml:92]: super NIT: use `oc get node` for consistency

File: bindata/network/openshift-sdn/sdn.yaml [hunk 4]
```diff
@@ -66,7 +66,56 @@ spec:
...
+                  echo "create hostsubnet CR for this node"
```
Comment (kyrtapz) [inline bindata/network/openshift-sdn/sdn.yaml:93]: NIT Add OVN_SUBNET and NODE_IP values to the log

File: bindata/network/openshift-sdn/sdn.yaml [hunk 5]
```diff
@@ -66,7 +66,56 @@ spec:
...
+          subnet: ${SUBNET}
+          EOF
+              else
+                  echo "cannot discover host subnet"
+                  exit 1
+              fi
+          fi
+          if ovs-vsctl br-exists br-ex; then
+            echo "br-ex exists, sleep..."
+            sleep infinity
```
Comment (kyrtapz) [inline bindata/network/openshift-sdn/sdn.yaml:112]: To avoid delays in killing the container we could use the following instead:
```
trap : TERM INT; sleep infinity & wait
```

Additionally would it be good to put an exit after the sleep? not sure about the conventions/best practice here.

File: bindata/network/ovn-kubernetes/common/002-rbac-node.yaml [hunk 6]
```diff
@@ -79,6 +79,7 @@ rules:
   - get
   - list
   - watch
+  - patch
```
Comment (kyrtapz) [inline bindata/network/ovn-kubernetes/common/002-rbac-node.yaml:82]: This should be under `{{- if .IsNetworkTypeMigration }}` condition.
Is patch required for namespaces, nodes and pods?

File: bindata/network/ovn-kubernetes/common/002-rbac-node.yaml [hunk 7]
```diff
@@ -184,7 +185,11 @@ rules:
   - get
   - list
   - watch
-
+{{- if .IsNetworkTypeMigration }}
+- apiGroups: ['network.openshift.io']
+  resources: ['hostsubnets']
+  verbs: ['get', 'list', 'create', 'patch']
```
Comment (kyrtapz) [inline bindata/network/ovn-kubernetes/common/002-rbac-node.yaml:191]: Do we need create and patch for hostsubnets in ovnkube-node?

File: bindata/network/ovn-kubernetes/self-hosted/common/004-config.yaml [hunk 8]
```diff
@@ -26,6 +26,9 @@ data:
 {{- if .OVNHybridOverlayEnable }}
     no-hostsubnet-nodes="kubernetes.io/os=windows"
 {{- end  }}
+{{- if .IsNetworkTypeMigration }}
+    no-hostsubnet-nodes="migration.network.openshift.io/plugin="
```
Comment (kyrtapz) [inline bindata/network/ovn-kubernetes/self-hosted/common/004-config.yaml:30]: What if OVNHybridOverlayEnable is also true?
Is no-hostsubnet-nodes additive?

File: bindata/network/ovn-kubernetes/common/002-rbac-node.yaml [hunk 9]
```diff
@@ -209,7 +214,21 @@ subjects:
   name: ovn-kubernetes-node
   namespace: openshift-ovn-kubernetes
 {{ end }}
-
+{{- if .IsNetworkTypeMigration }}
```
Comment (kyrtapz) [inline bindata/network/ovn-kubernetes/common/002-rbac-node.yaml:217]: Why does the `ovn-kubernetes-node` service account require the whole `openshift-ovn-kubernetes-node-limited` ClusterRole during the migration? I think we should limit it to what is required.
Additionally this binding already exists here: https://github.com/openshift/cluster-network-operator/blob/6569070efa9b09925568d0f321e59f19cb0ff563/bindata/network/ovn-kubernetes/common/002-rbac-node.yaml#L194-L216 (You would need to use the per-node certificate if the webhook is enabled)

File: bindata/network/ovn-kubernetes/self-hosted/common/004-config.yaml [hunk 10]
```diff
@@ -67,7 +70,7 @@ data:
     logfile-maxsize=100
     logfile-maxbackups=5
     logfile-maxage=0
-{{- if .OVNHybridOverlayEnable }}
+{{- if or .OVNHybridOverlayEnable }}
```
Comment (kyrtapz) [inline bindata/network/ovn-kubernetes/self-hosted/common/004-config.yaml:73]: Is this correct or are you missing a second operand?

**Review by kyrtapz — COMMENTED** [review]
Wen through the manifests.
I will go through the code changes next.

File: bindata/network/multus/multus.yaml [hunk 11]
```diff
@@ -135,14 +135,66 @@ data:
...
         "readinessindicatorfile": "/host/run/multus/cni/net.d/80-openshift-network.conf",
 {{- else if eq .DefaultNetworkType "OVNKubernetes"}}
         "readinessindicatorfile": "/host/run/multus/cni/net.d/10-ovn-kubernetes.conf",
+{{- end}}
 {{- end}}
         "daemonSocketDir": "/run/multus/socket",
         "socketDir": "/host{{ .MultusSocketParentDir }}/socket"
     }
+{{- if .IsNetworkTypeMigration}}
+  daemon-config-lm-ovn.json: |
```
[the remark this answers is not in the document]
Comment (pliurh) [reply bindata/network/multus/multus.yaml:149]: There's a parameter `readinessindicatorfile` in Multus. Multus will watch the `readinessindicatorfile` to check if the CNI is ready on a node. During the migration, when nodes are running with different CNI plugins. We cannot use a static `readinessindicatorfile` for all the nodes. So here we render 2 different config files for Multus. Multus can choose the right config file based on the presence of br-ex on a node.

File: bindata/network/multus/multus.yaml [hunk 12]
```diff
@@ -135,14 +135,66 @@ data:
         "multusAutoconfigDir": "/host/run/multus/cni/net.d",
         "namespaceIsolation": true,
         "globalNamespaces": "default,openshift-multus,openshift-sriov-network-operator",
+{{- if not .IsNetworkTypeMigration}}
```
[the remark this answers is not in the document]
Comment (pliurh) [reply bindata/network/multus/multus.yaml:138]: The `readinessindicatorfile` parameter will be set during live migration in the config files `daemon-config-lm-ovn.json` and `daemon-config-lm-sdn.json` instead.

File: bindata/network/multus/multus.yaml [hunk 13]
```diff
@@ -135,14 +135,66 @@ data:
...
+        "cniConfigDir": "/host/etc/cni/net.d",
+        "multusConfigFile": "auto",
+        "multusAutoconfigDir": "/host/run/multus/cni/net.d",
+        "namespaceIsolation": true,
+        "globalNamespaces": "default,openshift-multus,openshift-sriov-network-operator",
+        "readinessindicatorfile": "/host/run/multus/cni/net.d/10-ovn-kubernetes.conf",
+        "daemonSocketDir": "/run/multus/socket",
+        "socketDir": "/host{{ .MultusSocketParentDir }}/socket"
+    }
+  daemon-config-lm-sdn.json: |
```
[the remark this answers is not in the document]
Comment (pliurh) [reply bindata/network/multus/multus.yaml:173]: As we want Multus to select the `readinessindicatorfile` based on the running CNI of the node, I cannot think of any better alternative here.

File: bindata/network/node-identity/self-hosted/node-identity.yaml [hunk 14]
```diff
@@ -45,10 +45,7 @@ spec:
               set +o allexport
             fi
 
-            ho_enable=
```
[the remark this answers is not in the document]
Comment (pliurh) [reply bindata/network/node-identity/self-hosted/node-identity.yaml:48]: OVN-K will try to remove hybrid overlay node annotations even when the hybrid overlay is not enabled. https://github.com/ovn-org/ovn-kubernetes/blob/ac6820df0b338a246f10f412cd5ec903bd234694/go-controller/pkg/ovn/master.go#L791

File: bindata/network/ovn-kubernetes/common/002-rbac-node.yaml (hunk 7 above)
Comment (pliurh) [reply bindata/network/ovn-kubernetes/common/002-rbac-node.yaml:191]: Yes, when rolling back, we need to create hostsubnet CRs in ovnkube-node.

File: bindata/network/ovn-kubernetes/self-hosted/common/004-config.yaml (hunk 8 above)
Comment (pliurh) [reply bindata/network/ovn-kubernetes/self-hosted/common/004-config.yaml:30]: We don't support rollback a HO-enabled cluster. Since OpenShiftSDN doesn't support HO. So these conditions cannot be true at the same time.

File: bindata/network/ovn-kubernetes/self-hosted/common/004-config.yaml [hunk 15]
```diff
@@ -78,7 +81,12 @@ data:
     hybrid-overlay-vxlan-port="{{.OVNHybridOverlayVXLANPort}}"
     {{- end }}
 {{- end  }}
-{{- if .IsSNO }}
+{{- if .IsNetworkTypeMigration }}
```
[the remark this answers is not in the document]
Comment (pliurh) [reply bindata/network/ovn-kubernetes/self-hosted/common/004-config.yaml:84]: I will move this to the same place as `OVNHybridOverlayEnable`.

File: bindata/network/ovn-kubernetes/self-hosted/multi-zone-interconnect/ovnkube-node.yaml [hunk 16]
```diff
@@ -130,6 +130,14 @@ spec:
         - -c
         - |
           set -e
+{{- if .IsNetworkTypeMigration }}
```
[the remark this answers is not in the document]
Comment (pliurh) [reply bindata/network/ovn-kubernetes/self-hosted/multi-zone-interconnect/ovnkube-node.yaml:133]: Yeah, this was written before `ovnkube-lib.sh`. I will move it to `ovnkube-lib.sh`.

File: bindata/network/ovn-kubernetes/self-hosted/multi-zone-interconnect/ovnkube-node.yaml [hunk 17]
```diff
@@ -361,6 +412,12 @@ spec:
               - -c
               - |
                 set -x
+{{- if .IsNetworkTypeMigration }}
+                if ! ip a|grep br-ex; then
+                  exit 0
+                fi
+{{- end }}
+                rm -f /var/run/ovn/ovnsb_db.pid
```
[the remark this answers is not in the document]
Comment (pliurh) [reply bindata/network/ovn-kubernetes/self-hosted/multi-zone-interconnect/ovnkube-node.yaml:420]: Maybe I can remove this line.

File: bindata/network/ovn-kubernetes/self-hosted/multi-zone-interconnect/ovnkube-node.yaml [hunk 18]
```diff
@@ -549,13 +680,21 @@ spec:
             (( retries += 1 ))
             if [[ "${retries}" -gt ${tries} ]]; then
               echo "$(date -Iseconds) - ERROR - ${key_cert} not found"
-              return 1
+              exit 1
```
[the remark this answers is not in the document]
Comment (pliurh) [reply bindata/network/ovn-kubernetes/self-hosted/multi-zone-interconnect/ovnkube-node.yaml:683]: Yeah. `return` shall only be used in a function.

File: pkg/controller/operconfig/operconfig_controller.go [hunk 19]
```diff
@@ -252,15 +252,25 @@ func (r *ReconcileOperConfig) Reconcile(ctx context.Context, request reconcile.R
...
-			fmt.Sprintf("Internal error while merging cluster configuration and operator configuration: %v", err))
+	clusterConfig := &configv1.Network{}
+	if err := r.client.Default().CRClient().Get(ctx, types.NamespacedName{Name: names.CLUSTER_CONFIG}, clusterConfig); err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
 		return reconcile.Result{}, err
 	}
 
+	if !meta.IsStatusConditionPresentAndEqual(clusterConfig.Status.Conditions, names.NetworkTypeMigrationInProgress, metav1.ConditionTrue) {
```
Comment (kyrtapz) [inline pkg/controller/operconfig/operconfig_controller.go:263]: Please add a comment on why we don't want to merge in cluster config during live migration.
Considering that Conditions do not exist in the default API today wouldn't that `if` always result in false?

We need to add the featuregate checks for the live migration so we don't affect the default flow for now.

File: pkg/controller/operconfig/operconfig_controller.go [hunk 20]
```diff
@@ -302,20 +312,21 @@ func (r *ReconcileOperConfig) Reconcile(ctx context.Context, request reconcile.R
 		log.Printf("Using detected MTU %d", mtu)
 	}
 
+	newOperConfig := operConfig.DeepCopy()
```
Comment (kyrtapz) [inline pkg/controller/operconfig/operconfig_controller.go:315]: What is the reason for moving the copying here?

File: pkg/names/names.go [hunk 21]
```diff
@@ -134,6 +134,17 @@ const OpenShiftComponent = "openshift.io/owning-component"
 // ClusterNetworkOperatorJiraComponent is the jira component name for the cluster-network-operator
 const ClusterNetworkOperatorJiraComponent = "Networking / cluster-network-operator"
 
+// Status condition types of network.config for live migration
+const (
+	NetworkTypeMigrationAnnotation         string = "network.openshift.io/live-migration"
```
Comment (kyrtapz) [inline pkg/names/names.go:139]: This is not a condition, let's move it out from this block. 
Additionally please add a doc to each variable.

File: pkg/controller/operconfig/operconfig_controller.go [hunk 22]
```diff
@@ -345,6 +354,7 @@ func (r *ReconcileOperConfig) Reconcile(ctx context.Context, request reconcile.R
 	}
 
 	// once updated, use the new config
+	newOperConfig.Spec.Migration = operConfig.Spec.Migration
```
Comment (kyrtapz) [inline pkg/controller/operconfig/operconfig_controller.go:357]: Why is this required?
Please add a comment.

File: pkg/network/multus.go [hunk 23]
```diff
@@ -37,7 +37,11 @@ func renderMultus(conf *operv1.NetworkSpec, bootstrapResult *bootstrap.Bootstrap
 	usedhcp, usewhereabouts := detectAuxiliaryIPAM(conf)
 	h := bootstrapResult.Infra.APIServers[bootstrap.APIServerDefault].Host
 	p := bootstrapResult.Infra.APIServers[bootstrap.APIServerDefault].Port
-	objs, err = renderMultusConfig(manifestDir, string(conf.DefaultNetwork.Type), usedhcp, usewhereabouts, h, p, bootstrapResult)
+	isNetworkTypeMigration := false
```
Comment (kyrtapz) [inline pkg/network/multus.go:40]:
```suggestion
	isNetworkTypeLiveMigration := false
```

File: pkg/network/render.go [hunk 24]
```diff
@@ -649,9 +670,9 @@ func isDefaultNetworkChangeSafe(prev, next *operv1.NetworkSpec) []error {
 }
 
 func isMigrationChangeSafe(prev, next *operv1.NetworkSpec) []error {
-	if prev.Migration != nil && next.Migration != nil && prev.Migration.NetworkType != next.Migration.NetworkType {
-		return []error{errors.Errorf("cannot change migration network type after migration has started")}
-	}
+	// if prev.Migration != nil && next.Migration != nil && prev.Migration.NetworkType != next.Migration.NetworkType && next.Migration.Mode != operv1.LiveNetworkMigrationMode {
```
Comment (kyrtapz) [inline pkg/network/render.go:673]: This check is valid, I don't think we should remove that.

**Review by kyrtapz — COMMENTED** [review]
Went through the code but will need a second pass.

File: pkg/controller/clusterconfig/migration.go [hunk 25]
```diff
@@ -0,0 +1,189 @@
...
+		return err
+	}
+	if currentOperConfig.Spec.Migration != nil && currentOperConfig.Spec.Migration.Mode != operv1.LiveNetworkMigrationMode {
+		// In offline migration, propagate the network type to the operator config
+		return nil
+	}
+
+	if _, ok := clusterConfig.Annotations[names.NetworkTypeMigrationAnnotation]; ok {
+		// In live migration, preserve the network type change
+		operConfig.Spec.DefaultNetwork.Type = currentOperConfig.Spec.DefaultNetwork.Type
```
[the remark this answers is not in the document]
Comment (pliurh) [reply pkg/controller/clusterconfig/migration.go:41]: `operConfig` is rendered based on the content `network.config` CR. During the migration, we want to switch the `DefaultNetwork.Type` of `network.operator` when the prerequisit steps are completed.

File: pkg/controller/clusterconfig/migration.go [hunk 26]
```diff
@@ -0,0 +1,189 @@
...
+		log.Printf("MCP is updating, so we don't modify the operator config")
+		return nil
+	}
+
+	mtuApplied := meta.IsStatusConditionPresentAndEqual(status.Conditions, names.NetworkTypeMigrationMTUReady, metav1.ConditionTrue)
+	cniReady := meta.IsStatusConditionPresentAndEqual(status.Conditions, names.NetworkTypeMigrationTargetCNIAvailable, metav1.ConditionTrue)
+	mcApplied := meta.IsStatusConditionPresentAndEqual(status.Conditions, names.NetworkTypeMigrationTargetCNIInUse, metav1.ConditionTrue)
+
+	var isRollback bool
+	if clusterConfig.Spec.NetworkType == "OpenShiftSDN" {
```
[the remark this answers is not in the document]
Comment (pliurh) [reply pkg/controller/clusterconfig/migration.go:110]: The main difference is when we apply the routeble MTU. For migration, routable MTU is applied before switching the nodes CNI to OVN-K. For rollback, the routable MTU is applied togther with switching the nodes CNI to OVN-K. When a node is rebooted and SDN is used as the node CNI. So that pods on SDN nodes wouldn't send packets with MTU larger than that cannot be handled by OVN-K.

File: pkg/controller/clusterconfig/migration.go [hunk 27]
```diff
@@ -0,0 +1,189 @@
...
+	return -1, nil
+}
+
+func (r *ReconcileClusterConfig) prepareOperatorConfigForNetworkTypeMigration(ctx context.Context, clusterConfig *configv1.Network, operConfig *operv1.Network) error {
+	status := clusterConfig.Status
+	if status.Conditions == nil {
+		return fmt.Errorf("status.Conditions is not initialized")
+	}
+
+	mcpCondition := meta.FindStatusCondition(status.Conditions, names.NetworkTypeMigrationMTUReady)
```
[the remark this answers is not in the document]
Comment (pliurh) [reply pkg/controller/clusterconfig/migration.go:96]: This condition name is for user facing. Under the hood, we're checking the MCP condtions here.

File: bindata/network/openshift-sdn/sdn.yaml (hunk 2 above)
Comment (pliurh) [reply bindata/network/openshift-sdn/sdn.yaml:74]: The `jq` command is not available in the SDN image. I don't think we can support rollback with multiple node-subnets.

File: pkg/controller/clusterconfig/migration.go [hunk 28]
```diff
@@ -0,0 +1,189 @@
...
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+
+	"github.com/openshift/cluster-network-operator/pkg/names"
+	
+)
+
+func (r *ReconcileClusterConfig) processNetworkTypeMigration(ctx context.Context, request reconcile.Request, clusterConfig *configv1.Network, operConfig *operv1.Network) error {
+	var err error
+	currentOperConfig := &operv1.Network{}
+	err = r.client.Default().CRClient().Get(ctx, types.NamespacedName{Name: names.OPERATOR_CONFIG}, currentOperConfig)
```
[the remark this answers is not in the document]
Comment (pliurh) [reply pkg/controller/clusterconfig/migration.go:27]: I added a check before invoking `processNetworkTypeMigration`

File: bindata/network/multus/multus.yaml [hunk 29]
```diff
@@ -135,14 +135,66 @@ data:
         "multusAutoconfigDir": "/host/run/multus/cni/net.d",
         "namespaceIsolation": true,
         "globalNamespaces": "default,openshift-multus,openshift-sriov-network-operator",
+{{- if not .IsNetworkTypeLiveMigration}}
```
Comment (kyrtapz) [inline bindata/network/multus/multus.yaml:138]: Is `daemon-config.json` used during migration? If not I don't see why this is needed

File: bindata/network/node-identity/self-hosted/node-identity.yaml [hunk 30]
```diff
@@ -115,7 +112,7 @@ spec:
             name: ovnkube-identity-cm
       terminationGracePeriodSeconds: {{.NetworkNodeIdentityTerminationDurationSeconds}}
       nodeSelector:
-        node-role.kubernetes.io/master: ""
+        # node-role.kubernetes.io/master: ""
```
[the remark this answers is not in the document]
Comment (kyrtapz) [reply bindata/network/node-identity/self-hosted/node-identity.yaml:115]: the webhook has to be colocated with the API servers so I don't see a point in removing that.

File: bindata/network/node-identity/self-hosted/node-identity.yaml (hunk 14 above)
[the remark this answers is not in the document]
Comment (kyrtapz) [reply bindata/network/node-identity/self-hosted/node-identity.yaml:48]: thanks, can you add a comment please

File: pkg/controller/operconfig/migration.go [hunk 31]
```diff
@@ -616,3 +639,322 @@ func unstructuredEgressIpObject(egressIpName string, egressIps []interface{}, ne
...
+				if err := r.client.Default().CRClient().Update(ctx, &node); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (r *ReconcileOperConfig) ensureMachineConfigPools(ctx context.Context, clusterConfig *configv1.Network, condType string, isMatch match, nowTimestamp metav1.Time) (bool, bool, error) {
```
Comment (kyrtapz) [inline pkg/controller/operconfig/migration.go:869]: NIT: Please add a doc text describing the returned arguments

**Comment by kyrtapz** [comment]
hypershift failure looks unrelated, it is currently across all PRs:
```
Failed to generate payload: error getting ignition payload: failed to execute cluster-config-operator: cluster-config-operator process failed: /payloads/get-payload203317845/bin/cluster-config-operator: /lib64/libc.so.6: version `GLIBC_2.32' not found (required by /payloads/get-payload203317845/bin/cluster-config-operator)
      /payloads/get-payload203317845/bin/cluster-config-operator: /lib64/libc.so.6: version `GLIBC_2.34' not found (required by /payloads/get-payload203317845/bin/cluster-config-operator
```
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-network-operator/2091/pull-ci-openshift-cluster-network-operator-master-e2e-hypershift-ovn/1731985523595022336/artifacts/e2e-hypershift-ovn/run-e2e/artifacts/TestCreateCluster/namespaces/e2e-clusters-cvf29/hypershift.openshift.io/nodepools/example-45s4g-us-east-1a.yaml

thread: https://redhat-internal.slack.com/archives/C01C8502FMM/p1701716323597559

**Review by trozet — APPROVED** [review]
/approve

**Comment by martinkennelly** [comment]
Belated /lgtm :)
